presto-main/src/main/java/com/facebook/presto/operator/HashAggregationOperator.java

```diff
@@ -454,6 +494,52 @@ public HashAggregationBuilder getAggregationBuilder()
         return aggregationBuilder;
     }
 
+    private void processInputPage(Page page)
+    {
+        if (!preGroupedHashStrategy.isPresent()) {
+            unfinishedWork = aggregationBuilder.processPage(page);
+            return;
+        }
+
+        // Running in segmented aggregation mode.
+        if (firstUnfinishedSegment == null) {
+            // If this is the first page, treat the first segment in this page as the current segment.
+            firstUnfinishedSegment = page.getRegion(0, 1);
+        }
+
+        Page pageOnPreGroupedChannels = page.extractChannels(preGroupedChannels);
+        int lastRowInPage = page.getPositionCount() - 1;
+        int lastSegmentStart = findLastSegmentStart(preGroupedHashStrategy.get(), pageOnPreGroupedChannels);
+        if (lastSegmentStart == 0) {
+            // The whole page is in one segment.
+            if (preGroupedHashStrategy.get().rowEqualsRow(0, firstUnfinishedSegment.extractChannels(preGroupedChannels), 0, pageOnPreGroupedChannels)) {
+                // All rows in this page belong to the same unfinished segment, process the whole page.
```

Comment:
Even if it's a new segment, we can still claim "All rows in this page belong to the same unfinished segment"
so let's change it to `belong to the previous unfinished segment`